lox_services/persistence/database/utils.py

```diff
@@ -141,3 +143,53 @@ def make_temporary_table(
     table_ref = client.get_table(table_ref)
     table_ref.expires = datetime.now(timezone.utc) + timedelta(hours=1)
     client.update_table(table_ref, ["expires"])  # API request
+
+
+def make_validate_country_code() -> Callable[[pd.DataFrame, str], None]:
+    """Namespace for all the valid country codes."""
+    valid_codes = {country.alpha_2 for country in pycountry.countries}
+
+    def inner_validate_country_code(
+        df: pd.DataFrame, country_code_col: Union[str, Sequence[str]]
+    ) -> None:
+        """
+        Validate whether all rows in a DataFrame contain a valid ISO-3166-2 country code.
+
+        In the future, this check can be expanded as a custom data quality check for dataframe schema
+        (pandera, Great Expectations).
+
+        Parameters
+        ----------
+        df : pd.DataFrame
+            DataFrame containing a column with ISO-3166-2 country codes.
+        country_code_col : str
+            A string containing target country code column
+
+        Returns
+        -------
+        bool
+            True if all rows contain a valid ISO-3166-2 country code, False otherwise.
+        """
+        if isinstance(country_code_col, str):
+            country_code_col = [country_code_col]
+
+        error_mask = (
+            df[country_code_col]
+            .apply(lambda col: (col.notna()) & ~(col.isin(valid_codes)))
+            .any(axis="columns")
+        )
+        if error_mask.any():
+            print(
+                df.loc[error_mask, country_code_col].pipe(
+                    tabulate, headers="keys", tablefmt="psql"
+                )
+            )
+            raise ValueError(
```

Comment:
Can we hanlde NaN values not to raise a ValueError as this is not always specified by DHL, UPS ..

Reply:
Empty values get evaluated as False due to AND condition.